etl2566 - handling offset range call failure

zliu41 opened the pull request:
If unable to fetch latest offset from leader (response==null, response.hasError(), or caught runtime exception), retry three times. If still unable, skip the corresponding topics, proceed with other topics and report job failure.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java:
```diff
@@ -176,13 +183,18 @@ private SimpleConsumer createConsumer(JobContext context, String broker) {
+        earliestOffsetResponse = getLatestOffsetResponse(consumer, earliestOffsetInfo, context);
+      }
       consumer.close();
+      if (earliestOffsetResponse == null) {
```
Why there is no warning generation for latestOffsetResponse?

zliu41 replied:
For method 'generateLogWarnForSkippedTopics()', it doesn't matter whether we pass 'earliestOffsetInfo' or 'latestOffsetInfo' to it. Because it just needs to print the skipped topics, and 'earliestOffsetInfo' and 'latestOffsetInfo' contain exactly the same topics.

HenryCaiHaiying replied:
But it can still happen that earliestOffsetResponse != null but latestOffsetResponse == null since you are making two different kafka calls at different times, I think you need to check for both.

zliu41 replied:
In the 'if' block in line 188-190, it retrieves earliest only if latest!=null. So if latest=null, earliest will be null.

HenryCaiHaiying replied:
But if latest == null, you don't have earliestOffsetInfo, then what are you going to log in line 193?

zliu41 replied:
if latest==null, I don't have earliestOffsetResponse, but I do have earliestOffsetInfo, which is set in the for loop (181-184)

HenryCaiHaiying replied:
Looks good then.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java:
```diff
@@ -200,6 +212,34 @@ private SimpleConsumer createConsumer(JobContext context, String broker) {
+          throw new RuntimeException();
+        }
+        return offsetResponse;
+      } catch (RuntimeException e) {
```
The Kafka exception is not RuntimeException

zliu41 replied:
Fixed.

HenryCaiHaiying commented on camus-etl-kafka/src/main/java/com/linkedin/camus/etl/kafka/mapred/EtlInputFormat.java:
```diff
@@ -200,6 +212,34 @@ private SimpleConsumer createConsumer(JobContext context, String broker) {
+        }
+        return offsetResponse;
+      } catch (RuntimeException e) {
+        log.warn("Fetching offset from leader " + consumer.host() + ":" + consumer.port()
```
Need to put a sleep during iterations.

zliu41 replied:
Fixed.

zliu41 commented:
Added unit test for offset range call failure:
- Use a flag in EtlInputFormat (useMockConsumerForUnitTest). If flag==true, it will use a mocked consumer, which throws exception when trying to retrieve offsets.

Also addressed other minor issues.

zliu41 commented:
Rewrote unit test using the new mock framework. Besides the original test that throws exception for offset range calls, added two new tests:
- offsetResponse always has error.
- offsetResponse has error for the first 3 times, but succeeds the last time.

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -151,7 +154,15 @@ public void runJob1() throws Exception {
   }
 
   private void setupJob1() {
```
Broke the original setupJob1() into multiple small methods, each creating one mock, so that they can be reused.

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -137,12 +139,13 @@ public void before() throws IOException, NoSuchFieldException, IllegalAccessExce
   public void after() throws IOException {
     // Delete all camus data
     folder.delete();
+    _mocks.clear();
```
_mocks.clear() is needed because _mocks is used by multiple tests.

HenryCaiHaiying replied:
Looks good.

HenryCaiHaiying replied:
Or the _mocks shouldn't be an instance variable.  Since we are not sharing mocks between different tests, it probably should be method level variable.

zliu41 replied:
Still keeping _mocks as instance variable. Otherwise, a lot of methods need to add a parameter, which is cumbersome.

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -193,25 +208,117 @@ private void setupJob1() {
+
+  private void mockSimpleConsumer(TopicMetadataResponse metadataResponse, OffsetResponse offsetResponse,
+      FetchResponse fetchResponse) {
+    _simpleConsumer = EasyMock.createNiceMock(SimpleConsumer.class);
```
changed createMock() to createNiceMock(), because the new code needs to call consumer.host() and consumer.port() in generating log messages, which are not defined for the mocked consumer.

HenryCaiHaiying replied:
I prefer the createMock because it is more strict.  Can we just add consumer.host(), consumer.port() in the mocked consumer object?

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -193,25 +208,117 @@ private void setupJob1() {
     _mocks.add(_simpleConsumer);
     EasyMock.expect(_simpleConsumer.send((TopicMetadataRequest)EasyMock.anyObject())).andReturn(metadataResponse).times(1);
-    EasyMock.expect(_simpleConsumer.getOffsetsBefore((OffsetRequest)EasyMock.anyObject())).andReturn(offsetResponse).times(2);
+    EasyMock.expect(_simpleConsumer.getOffsetsBefore((OffsetRequest)EasyMock.anyObject())).andReturn(offsetResponse).anyTimes();
```
changed times(2) to anyTimes(), because it needs to be called more than twice in the new code (max retry = 3)

HenryCaiHaiying replied:
Can the 'times' be an argument so each test can set exactly how many times it was expecting?

zliu41 replied:
For the same reason, I think it shouldn't be an argument. We can have different methods, e.g., mockSimpleConsumer(), mockSimpleConsumerThrowsExceptionForOffsetRangeCall(), etc.

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -170,17 +181,21 @@ private void setupJob1() {
     EasyMock.expect(offsetResponse.offsets(EasyMock.anyString(), EasyMock.anyInt())).andReturn(new long[]{myMessages.size()}).times(1);
     // The second call is getEarliestOffset, we set the value to 0
     EasyMock.expect(offsetResponse.offsets(EasyMock.anyString(), EasyMock.anyInt())).andReturn(new long[]{0}).times(1);
+    EasyMock.expect(offsetResponse.hasError()).andReturn(false).anyTimes();
```
This is added because otherwise, hasError() will return true by default.

HenryCaiHaiying replied:
Why not passing the expected return value as an argument to mockOffsetResponse(), this way you can control whether the mock object should return true or false.

zliu41 replied:
Some tests may have more complex requirements, e.g., first two times has error, third time succeeds. Also, some tests may need to change the return value of a different method, rather than hasError(). 
So, I think there should be a separate method for each need. Currently there are 3: mockOffsetResponse(), mockOffsetResponseWithError() and mockOffsetResponseThridRetrySucceed().

zliu41 commented on camus-etl-kafka/src/test/java/com/linkedin/camus/etl/kafka/CamusJobTestWithMock.java:
```diff
@@ -193,25 +208,117 @@ private void setupJob1() {
+  }
+
+  @Test
+  public void testJobOffsetRangeCallThirdRetrySucceed() throws Exception {
```
this test doesn't have verify() or assert(), because it passes if the job succeeds (after three retries).

HenryCaiHaiying replied:
Even for three retries, can we calculate how many times the mock will be called for verify() to pass?

zliu41 commented:
Regression test passed on Nertz.

zliu41 commented:
- Replaced all nice mocks with regular mocks
- use anyTimes() as few as possible.
- no longer verify exception messages.
